source/adapters/opencl/adapter.cpp

```diff
@@ -8,16 +8,49 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include "adapter.hpp"
 #include "common.hpp"
-#include "logger/ur_logger.hpp"
+#include "ur/ur.hpp"
+
+#ifdef _MSC_VER
+#include <Windows.h>
+#else
+#include <dlfcn.h>
+#endif
+
+ur_adapter_handle_t_::ur_adapter_handle_t_() {
+#ifdef _MSC_VER
+  // Loading OpenCL.dll increments the libraries internal reference count.
+  auto handle = LoadLibraryA("OpenCL.dll");
+
+#define CL_CORE_FUNCTION(FUNC)                                                 \
+  FUNC = reinterpret_cast<decltype(::FUNC) *>(GetProcAddress(handle, "FUNC"));
+#include "core_functions.def"
+#undef CL_CORE_FUNCTION
+
+  // So we can safely decrement it here wihtout actually unloading OpenCL.dll.
+  FreeLibrary(handle);
+#else
+  // Loading libOpenCL.so to get the library handle but don't dlclose it as
+  // this causes a segfault when attempting to call any OpenCL entry point.
+  auto handle = dlopen("libOpenCL.so", RTLD_LOCAL);
```

Comment:
I have the same concern here that I have on `LoadLibraryA` on loading the same opencl.dll used.

Reply:
The way we link `libOpenCL.so` is already making use of the standard dynamic library loading used by `dlopen()`. We don't link to an absolute path but rather the `libOpenCL.so.1` name and that's used for the library lookup. For example, on my system it's found `/lib/x86_64-linux-gnu/libOpenCL.so.1`.

```
$ ldd ~build/install/lib/libur_adapter_opencl.so
...
        libOpenCL.so.1 => /lib/x86_64-linux-gnu/libOpenCL.so.1 (0x00007b389a819000)
...
```

But if I set `LD_LIBRARY_PATH` it will use that.

```
$ LD_LIBRARY_PATH=$Projects/KhronosGroup/OpenCL-ICD-Loader/build/bin \
    ldd ~build/install/lib/libur_adapter_opencl.so
...
        libOpenCL.so.1 => /home/benie/Projects/KhronosGroup/OpenCL-ICD-Loader/build/bin/libOpenCL.so.1 (0x0000771087e10000)
...
```

My point being, the look up for done by `ld` during initialization is exacrlt the same as the lookup done by `dlopen()`. Not only that but `dlopen()` won't load the library again if its already present.

But just to be sure, lets invoke @ldrumm to confirm or deny my understanding.

Reply:
man dlopen:

>   If  the  same shared object is opened again with dlopen(), the same object handle is returned.

Reply:
make sense. Thanks for clarification!